[리액트 7기 표후동] 레스토랑 로그인 페이지 및 리뷰 작성 페이지 구현하기

ndopy opened the pull request:
이전 강의들 복습 및 반복숙달하면서 붙여나가겠습니다. 👨‍🚀

hannut91 committed:
Move index.html file to public directory

Commit:
Failing test : renders title

- LoginPage renders title : "Login"

Commit:
Revert "Failing test : renders title"

This reverts commit e2d4ed82e3dba880677fded57be5bbf8e4d5bb76.

Commit:
Add failing test : <LoginPage />

- it renders title : "Login"

Commit:
Pass failing test : <LoginPage />

- it renders title : "Login"

Commit:
Refactor test : <LoginPage />

- Add "import LoginPage from './LoginPage';"

Commit:
Add failing test : <LoginFormContainer />

- it renders input controls

Commit:
Pass failing test : <LoginFormContainer />

- it renders input controls

Commit:
Add failing test : <LoginPage />

- it renders input controls

Commit:
Pass failing test : <LoginPage />

- it renders input controls

Commit:
Add failing test : <LoginFormContainer />

- it renders login button

Commit:
Pass failing test : <LoginFormContainer />

- it renders login button

Commit:
Add failing test : <LoginForm />

- renders input controls

Commit:
Pass failing test : <LoginForm />

- renders input controls

Commit:
Add failing test : <LoginForm />

- renders login button
  - login button click

Commit:
Add function <LoginFormContainer />

- remove LoginForm elements
- create function handleSubmit
- add handleSubmit in <LoginForm/> as onSubmit

Commit:
Pass failing test : <LoginForm />

- set parameter : onSubmit
- implement function handleClick

Commit:
Add failing test: <LoginForm />

- renders input controls "and change events"

Commit:
Pass failing test: <LoginForm />

- renders input controls "and change events"

moonkii commented on src/LoginFormContainer.jsx:
```diff
@@ -0,0 +1,22 @@
+    <>
+      <div>
+        <label htmlFor="login-email">email</label>
+        <input
+          type="email"
+          placeholder="email"
+          id="login-email"
+        />
+      </div>
+      <div>
+        <label htmlFor="login-password">password</label>
+        <input
+          type="password"
+          placeholder="password"
+          id="login-password"
+        />
+      </div>
```
의미있는 마크업까지는 잘 하셨는데, 항상 완성된 형태를 유지해주시면 좋아요.
나머지 속성들도 채워서 만들어 줄 수 있을거에요.

moonkii commented on src/LoginFormContainer.test.jsx:
```diff
@@ -0,0 +1,16 @@
+  );
+
+  it('renders input controls', () => {
+    const { getByLabelText } = renderLoginFormContainer();
+
+    expect(getByLabelText('email')).not.toBeNull();
+    expect(getByLabelText('password')).not.toBeNull();
```
`getBy~` 와 `queryBy~` 를 구분해서 사용해주시나요?

moonkii commented on src/LoginPage.jsx:
```diff
@@ -0,0 +1,10 @@
+  return (
+    <div>
+      <h2>Login</h2>
+      <LoginFormContainer />
```
지금까지 구조에서는 미리 구조를 설계하여 만드신 것 같은 느낌이 들어요.
만약에 저라면 이미 `LoginPage` 에서 완성된 형태를 만든 후 리펙터링하면서 컴포넌트를  분리해줄 것 같아요.

ndopy replied:
"우선은 최대한 강의대로 해보자" 라고 생각해서 따라하다보니 이렇게 됐네요.

말씀해주신 내용이  "`<LoginPage />` 에 코드를 완성시키고,
그 이후에 Container 나 Page 등으로 분리하면서 리팩토링" 하라는 의미이신 것 같아요.

강의 이후에 반복해서 구현할 때는 말씀해주신 방향으로 진행해보겠습니다.

moonkii replied:
강의에서도 완성된 코드가 중요한게 아니라 오히려 완성해 나가는 흐름이 더 중요해요.
코드 패턴을 익히는것도 도움이 되지만 강의를 보시면서 어떤 사고의 흐름이나 과정으로 코드를 완성해 나가는지 유심히 보시면 좋아요.

moonkii reviewed (commented):
![Uploading image.png…]()

TDD 를 할 때 테스트 코드를 먼저 작성하신 건 잘 하신 것 같은데 미리 구조를 정하고 시작하기 보다는 완성된 상태에서 리펙터링해 나가면서 좋은 구조를 '발견'해 나가신다는걸 기억해주시면 좋아요.
그래야 TDD를 하는 의미가 있는거니까요.

Commit:
Refactor <LoginForm /> test

- refactor code for multiple inputs test with array and forEach

Commit:
Add updateLoginField Action in <LoginFormContainer />

Commit:
Add useSelector in <LoginFormContainer /> test

Commit:
Add useSelector in <LoginFormContainer />

Commit:
Add beforeEach in <LoginPage /> test

- add mockImplementation for useSelector
  - loginFields

Commit:
Refactor <LoginPage /> test

- getByText -> queryByText is suitable for a null check test

Commit:
Refactor <LoginFormContainer /> test

- getByText -> queryByText is suitable for a null check test

Commit:
Update <LoginFormContainer /> test

- email, password input 값에 대한 테스트로 업데이트한다.

- 이 때는 queryByText -> getByText 가 더 적절할 것 같다.
  - value 를 얻으려면 element 가 존재한다는 전제가 있어야하기 때문에?

Commit:
Refactor <LoginForm /> test

- <LoginForm /> 에 loginFields props 를 전달한다.
- input 의 onChange 이전에 사용할 email, password 값을 설정한다. (previousValue)

Commit:
Refactor <LoginForm /> test

- 공통으로 사용되는 코드를 분리한다.

Commit:
Fix <LoginForm /> test

- 오탈자 수정 : loginFiels -> loginFields

Commit:
Add loginFields in <LoginForm />

- loginFields 상태를 받아서 email, password input 의 값으로 설정한다.

Commit:
Fix <LoginFormContainer />

- getBy"Label"Text 로 input 의 value 를 확인해야하는데 실수로 getByText 라고 입력했다.
- getByText -> getByLabelText

Commit:
Add failing test : updateLoginField action in Reducer

- updateLoginField : email, password 상태 값을 사용자가 입력한 값으로 업데이트해야 한다.
  -> email input 에 'test@test' 라고 입력하면, state 의 email 값도 'test@test' 가 되어야 한다.

Commit:
Pass failing test : updateLoginField action in Reducer

- reducer 의 initialState 에 loginFields 가 추가되어서 "when previous state is undefined" 테스트에서도 에러가 발생했다.
  - 테스트 코드의 initialState 에도 loginFields 를 추가해서 초기 상태를 같게 해줘야 한다.

Commit:
Refactor failing test : updateLoginField action in Reducer

- 이메일, 비밀번호가 변경되는 경우를 나눈다.
  - when email changes
  - when password changes

Commit:
Pass failing test : updateLoginField action in Reducer

- loginFields 에 기존 상태 값을 추가한다.
  - ...state.loginFields

Commit:
Add actions : requestLogin, setAccessToken

- requestLogin : API 서버에 로그인을 요청한다.
  ㄴ 로그인 정보로 email, password 를 전달한다.
  ㄴ 로그인이 성공하면 액세스 토큰을 얻는다.
  ㄴ 액세스 토큰(로그인 정보)을 저장한다.

- setAccessToken : 액세스 토큰을 저장한다.
  ㄴ 로그인이 성공하여 얻은 액세스 토큰을 저장한다.

Commit:
Add failing test : setAccessToken by reducer

- 초기 상태
  ㄴ accessToken 이 없다.

- 리듀서가 setAccessToken 액션을 실행하면
  ㄴ 'token' 이라는 액세스 토큰을 얻었다면 state 의 accessToken 값도 'token' 이 되어야 한다.

Commit:
Add function in api : postLogin

- API 서버에서 로그인을 요청하는 동작을 담당한다.
  ㄴ 실제 동작 구현 필요

Commit:
Pass failing test : setAccessToken by reducer

- initialState 에 accessToken 의 기본값을 추가한다.
- accessToken 을 전달받아서 state 에 accessToken 을 저장하는 액션을 구현한다.

- 'when previous state is undefined' 테스트의 initialState 에도 accessToken 의 기본값인 빈 값('')을 추가한다.

Commit:
Add accessToken in <LoginFormContainer />

- 로그인 요청 후 얻을 액세스 토큰을 화면에 표시한다.

Commit:
Add path for LoginPage

- 로그인 페이지를 연결한다.

Commit:
Fix initialState of reducer

- loginFields 상태의 기본값에 email, password 프로퍼티를 추가한다.

Commit:
Update function postLogin in api

- api 서버에 로그인을 요청하고, 로그인에 성공하면 액세스 토큰을 얻는 동작을 구현한다.
- 이것을 간편하게 해주는 axios 라는 라이브러리가 있다고 한다.

Commit:
Fix function requestLogin

- await 키워드가 빠져있어서 추가한다.

Commit:
Update browserslist version

-다음 명령어를 실행했다.
  ㄴ "npx browserslist@latest --update-db"

Commit:
Add failing test : <RestaurantPage />

Commit:
Add failing test : <RestaurantContainer />

Commit:
Pass failing test : <RestaurantContainer />

- <RestaurantContainer/> 에 <ReviewForm /> 을 추가할 것이다.
  - 현재는 간단하게 function 으로 구현했지만 이후에 컴포넌트로 생성해서 분리하자.

Commit:
Update test : <RestaurantContainer />

- 테스트 이름 변경 : renders review write form
- 평점과 리뷰 내용 입력에 대한 onChange 관련 테스트를 추가한다.

Commit:
Add review description in <RestaurantContainer />

Commit:
Update failing test : <RestaurantContainer />

- 평점과 리뷰 내용을 입력하면 해당 값을 업데이트하는 액션이 dispatch 되어야 한다.

Commit:
Pass failing test : <RestaurantContainer />

- 평점과 리뷰 내용을 입력하면 해당 값을 업데이트하는 액션이 dispatch 되어야 한다.

- updateReviewField Action Creator 를 추가한다.

Commit:
Add failing test : Reducer

- 평점이 입력되어 값이 변경됐을 경우의 테스트
- 리뷰 내용이 입력되어 값이 변경됐을 경우의 테스트

Commit:
Refactor <RestaurantContainer />

- 평점, 리뷰 내용의 상태를 변경하려면 관리하는 상태에서의 이름과 같게 해야 한다.
- initialState 의 reviewFields 에는 평점과 리뷰 내용이 각각 score, description 이라는 프로퍼티로 작성했다.
  - '평점' input 의 name 을 변경 : score
  - '리뷰 내용' input 의 name 을 변경 : description

Commit:
Update initialState in reducer test

- 'when previous state is undefined' 테스트의 initialState 에 reviewFields 를 추가한다.

Commit:
Refactor test : <RestaurantContainer />

- 평점, 리뷰 내용 각 input 의 name 을 score, description 으로 수정하였으므로, 테스트에서도 이에 맞게 수정한다.

Commit:
Pass failing test : Reducer

- initialState 에 reviewFields 를 추가한다.
- updateReviewField 액션 동작을 구현한다.

moonkii commented on src/LoginForm.jsx:
```diff
@@ -0,0 +1,46 @@
+  return (
+    <>
+      <div>
+        <label htmlFor="login-email">email</label>
+        <input
+          type="email"
+          name="email"
+          placeholder="email"
+          id="login-email"
+          value={email}
+          onChange={handleChange}
+        />
```
이 부분을 따로 컴포넌트로 분리해봐도 좋을 것 같아요.

moonkii commented:
가능하면 강의를 보면서 코드를 따라하기 보다는 강의를 보면서 공부를하고, 코드는 최대한 스스로 작성해볼려고 노력해보세요.
공부효과가 더 클거에요.

Commit:
Refactor test : <RestaurantContainer />

- score, description input 의 onChange 에 대한 테스트는 테스트 내용이 같다.
  - 따라서 두 input 을 controls 배열에 담아서 forEach 로 처리할 수 있다.

Commit:
Add failing test : <RestaurantForm />

- it renders <ReviewForm />

Commit:
Pass failing test : <RestaurantForm />

- it renders <ReviewForm />

- <RestaurantContainer /> 에 작성되어 있던 function ReviewForm 을 컴포넌트로 분리한다.

Commit:
Update test : <RestaurantForm />

- it renders review write fields

Commit:
Add test : <RestaurantForm />

- it listens input change event

Commit:
Add failing test : <TextField />

- it renders label and input control

Commit:
Pass failing test : <TextField />

- it renders label and input control

Commit:
Add failing test : <TextField />

- <TextField /> 의 input 에 type 을 전달해서 명시적으로 전달받지 않으면 기본값으로 "text" 를 사용하도록 테스트 한다.

Commit:
Pass failing test : <TextField />

- <TextField /> 의 input 에 type 을 전달해서 명시적으로 전달받지 않으면 기본값으로 "text" 를 사용하도록 테스트 한다.

Commit:
Refactor test : <TextField />

- <TextField /> 를 render 하는 코드가 반복해서 사용되므로 함수로 분리한다.
- 테스트를 세분화한다.
  - it renders label and input control
  - it renders "number" input control
  - it renders "text" input control

Commit:
Add test : <TextField />

- it listens input change event

Commit:
Refactor <ReviewForm>

- <TextField> 컴포넌트를 이용해서 리팩토링한다.

Commit:
Add failing test : <ReviewForm />

- it renders "Send" button

Commit:
Revert "Add failing test : <ReviewForm />"

This reverts commit 42b9e50dccd2dd59fd32e6b78e0d421a0dd7914b.

Commit:
Add failing test : <ReviewForm />

- it renders "Send" button

moonkii commented on src/LoginFormContainer.test.jsx:
```diff
@@ -0,0 +1,43 @@
+  );
+
+  it('renders input controls', () => {
+    const { getByLabelText } = renderLoginFormContainer();
+
+    expect(getByLabelText('email').value).toBe('test@test');
+    expect(getByLabelText('password').value).toBe('1234');
```
render 테스트에서는 queryBy~ 가 의미상으로 더 적절할거에요.

moonkii commented on src/ReviewForm.jsx:
```diff
@@ -0,0 +1,19 @@
+export default function ReviewForm({ onChange }) {
+  return (
+    <>
+      <TextField
+        label="평점"
+        type="number"
+        name="score"
+        onChange={onChange}
+      />
+      <TextField
+        label="리뷰 내용"
+        name="description"
+        onChange={onChange}
```
애매하게 미완성된 형태로 남아있는 것 같아요.
value 속성도 채워주시고 자체로도 이미 작동하는 형태가 되어야해요.
보통은 TDD가 제대로 안되는 경우 이런경우가 많아요. 과정을 많이 뛰어넘은거죠.
